src/cryptonote_core/service_node_voting.h

```diff
@@ -51,30 +51,51 @@ namespace service_nodes
 
   struct voter_to_signature
   {
+    uint8_t           version = 0;
     uint16_t          voter_index;
     crypto::signature signature;
 
     BEGIN_SERIALIZE()
-      FIELD(voter_index)
+      VARINT_FIELD(voter_index)
       FIELD(signature)
     END_SERIALIZE()
   };
 
   struct checkpoint_vote { crypto::hash block_hash; };
   struct deregister_vote { uint16_t worker_index; };
 
-  enum struct quorum_type
+  enum struct quorum_type : uint8_t
   {
-    deregister = 0,
+    uptime = 0,
     checkpointing,
     count,
+    invalid = count,
+  };
+
+  enum struct quorum_vote_type : uint8_t
+  {
+    uptime_deregister = 0,
+    checkpoint,
+    checkpoint_deregister,
```

Comment:
I don't think that we should have different types of deregister transactions: rather we should build the deregistration criteria into the deregistration quorum.

So, instead of "is uptime proofs more than 2 hours" we will have the criteria "is uptime proofs more than 2 hours OR were there checkpoints more than 2 hours old that did not have a checkpoint vote submitted".

I don't even think we have to worry about verifying the validity of those votes for the purpose of the deregistration vote.

Reply:
This became more apparent when I had a look at your crediting system. I agree, I want to funnel all the deregistering through your subsystem which means I need to rewrite that chunk of this.